### PR TITLE
Modernize ranges::copy_if

### PR DESCRIPTION
**Pull request by miscco**
Use the new way of delegating the work and improve the tests

### CODE REVIEW

**Comment by miscco** [comment]
>C:\agent\_work\1\s\tests\std\tests\P0896R4_ranges_alg_copy_if\test.cpp(84): fatal error C1060: compiler is out of heap space

Double ups

@BillyONeal we need MOOOOOOOOOORE RAM

File: stl/inc/algorithm [hunk 1]
```diff
@@ -1564,28 +1564,36 @@ namespace ranges {
...
+                                           _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            auto _First = _RANGES begin(_Range);
+            _Seek_wrapped(_First, _STD move(_Res.in));
+            return {_STD move(_First), _STD move(_Res.out)};
         }
         // clang-format on
+    private:
+        template <class _It, class _Se, class _Out, class _Pj, class _Pr>
+        _NODISCARD constexpr copy_if_result<_It, _Out> _Copy_if_unchecked(
+            _It _First, _Se _Last, _Out _Result, _Pr _Pred, _Pj _Proj) const {
```
Comment (CaseyCarter) [inline stl/inc/algorithm:1588]:
```suggestion
            _It _First, const _Se _Last, _Out _Result, _Pr _Pred, _Pj _Proj) const {
            _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It>);
            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
            _STL_INTERNAL_STATIC_ASSERT(indirect_unary_predicate<_Pr, projected<iterator_t<_Rng>, _Pj>>);
            _STL_INTERNAL_STATIC_ASSERT(indirectly_copyable<_It, _Out>);
            
```

File: stl/inc/algorithm [hunk 2]
```diff
@@ -1564,28 +1564,36 @@ namespace ranges {
...
+            auto _Res = _Copy_if_unchecked(_Ubegin(_Range), _Uend(_Range),
+                                           _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            auto _First = _RANGES begin(_Range);
+            _Seek_wrapped(_First, _STD move(_Res.in));
+            return {_STD move(_First), _STD move(_Res.out)};
         }
         // clang-format on
+    private:
+        template <class _It, class _Se, class _Out, class _Pj, class _Pr>
+        _NODISCARD constexpr copy_if_result<_It, _Out> _Copy_if_unchecked(
```
Comment (CaseyCarter) [inline stl/inc/algorithm:1587]:
```suggestion
        _NODISCARD static constexpr copy_if_result<_It, _Out> _Copy_if_unchecked(
```

File: stl/inc/algorithm [hunk 3]
```diff
@@ -1564,28 +1564,36 @@ namespace ranges {
...
+            _It _First, _Se _Last, _Out _Result, _Pr _Pred, _Pj _Proj) const {
+            for (; _First != _Last; ++_First) {
+                if (_STD invoke(_Pred, _STD invoke(_Proj, *_First))) {
+                    *_Result = *_First;
+                    ++_Result;
+                }
+            }
+            return {_STD move(_First), _STD move(_Result)};
```
Comment (CaseyCarter) [inline stl/inc/algorithm:1595]:
```suggestion

            return {_STD move(_First), _STD move(_Result)};
```

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp [hunk 4]
```diff
@@ -56,64 +55,53 @@ struct not_pair {
     }
 };
 
-constexpr void smoke_test() {
-    using ranges::copy_if, ranges::copy_if_result, ranges::iterator_t;
-    using std::same_as;
-    using P = not_pair<int, int>;
+using ranges::copy_if, ranges::copy_if_result, ranges::iterator_t;
+using std::same_as;
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp:59]: `using namespace std;` (and `const`-west) is idiomatic in modern Ranges tests.

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp [hunk 5]
```diff
@@ -56,64 +55,53 @@ struct not_pair {
...
+    static constexpr void call() {
+        { // Validate range overload
+            std::array<P, 2> output = {{{-1, -1}, {-1, -1}}};
+            In wrapped_input{input};
+            auto result = copy_if(wrapped_input, Write{output.data()}, is_odd, get_first);
+            STATIC_ASSERT(same_as<decltype(result), copy_if_result<iterator_t<In>, Write>>);
+            assert(result.in == wrapped_input.end());
+            if constexpr (std::equality_comparable<Write>) {
+                assert(result.out == Write{output.data() + 2});
+            }
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp:87]:
```suggestion
            assert(result.out.base() == output.data() + 2);
```
(And on 96-98.)

File: stl/inc/algorithm (hunk 2 above)
Comment (miscco) [reply stl/inc/algorithm:1587]: Will change but remove the const

File: stl/inc/algorithm [hunk 6]
```diff
@@ -1564,28 +1564,36 @@ namespace ranges {
...
-            for (; _UFirst != _ULast; ++_UFirst) {
-                if (_STD invoke(_Pred, _STD invoke(_Proj, *_UFirst))) {
-                    *_Result = *_UFirst;
-                    ++_Result;
-                }
-            }
-
-            _Seek_wrapped(_First, _STD move(_UFirst));
-            return {_STD move(_First), _STD move(_Result)};
+            auto _Res = _Copy_if_unchecked(_Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Last)),
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/algorithm]: Done

File: stl/inc/algorithm (hunk 2 above)
Comment (miscco) [reply stl/inc/algorithm]: Done

Comment (CaseyCarter) [reply stl/inc/algorithm]: ...and now I'm wondering if I missed this in the other reviews.

File: stl/inc/algorithm [hunk 7]
```diff
@@ -1564,28 +1564,41 @@ namespace ranges {
...
         template <input_range _Rng, weakly_incrementable _Out, class _Pj = identity,
             indirect_unary_predicate<projected<iterator_t<_Rng>, _Pj>> _Pr>
             requires indirectly_copyable<iterator_t<_Rng>, _Out>
         constexpr copy_if_result<borrowed_iterator_t<_Rng>, _Out> operator()(
             _Rng&& _Range, _Out _Result, _Pr _Pred, _Pj _Proj = {}) const {
-            return (*this)(
-                _RANGES begin(_Range), _RANGES end(_Range), _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            auto _UResult = _Copy_if_unchecked(
+                _Ubegin(_Range), _Uend(_Range), _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return {_Rewrap_iterator(_Range, _STD move(_UResult.in)), _STD move(_UResult.out)};
```
Comment (CaseyCarter) [inline stl/inc/algorithm:1580]: My mistake (I think) - we can't both `_Ubegin` and `_Rewrap_iterator` (which potentially calls `begin`) a single-pass range.
```suggestion
            auto _First = _RANGES begin(_Range);
            auto _UResult = _Copy_if_unchecked(
                _STD move(_First), _Uend(_Range), _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
            _Seek_wrapped(_First, _STD move(_UResult.in));
                        return {_STD move(_First), _STD move(_UResult.out)};
```

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp [hunk 8]
```diff
@@ -7,18 +7,18 @@
...
 
 #include <range_algorithm_support.hpp>
 
+using namespace std;
 template <class T, class U>
 struct not_pair {
     T first;
     U second;
 
     auto operator<=>(const not_pair&) const = default;
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp:19]: Pre-existing inconsistent `const`:
```suggestion
    auto operator<=>(not_pair const&) const = default;
```

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp [hunk 9]
```diff
@@ -36,84 +36,70 @@ struct not_pair {
...
+    ranges::copy_if_result<int*, int*>>);
 
 struct instantiator {
-    template <class In, class Out>
-    static void call(In&& in = {}, Out out = {}) {
-        using ranges::begin, ranges::copy_if, ranges::end, ranges::iterator_t;
+    static constexpr P input[3]    = {{1, 99}, {4, 98}, {5, 97}};
+    static constexpr P expected[2] = {{1, 99}, {5, 97}};
+
+    template <class Read, class Write>
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp:76]:
```suggestion
    template <ranges::input_range Read, indirectly_writable<ranges::range_reference_t<Read>> Write>
```

File: stl/inc/algorithm (hunk 7 above)
Comment (miscco) [reply stl/inc/algorithm:1580]: Ah we would have to call _Rewrap_iterator beforehand as we already iterated the range. Great catch!

File: stl/inc/algorithm [hunk 10]
```diff
@@ -1564,28 +1564,36 @@ namespace ranges {
...
             requires indirectly_copyable<iterator_t<_Rng>, _Out>
         constexpr copy_if_result<borrowed_iterator_t<_Rng>, _Out> operator()(
             _Rng&& _Range, _Out _Result, _Pr _Pred, _Pj _Proj = {}) const {
-            return (*this)(
-                _RANGES begin(_Range), _RANGES end(_Range), _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            auto _Res = _Copy_if_unchecked(_Ubegin(_Range), _Uend(_Range),
+                                           _STD move(_Result), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            auto _First = _RANGES begin(_Range);
+            _Seek_wrapped(_First, _STD move(_Res.in));
+            return {_STD move(_First), _STD move(_Res.out)};
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/algorithm]: As discussed this in not applicable for input ranges

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp (hunk 4 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp]: done

File: tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp [hunk 11]
```diff
@@ -7,18 +7,18 @@
 #include <compare>
 #include <concepts>
 #include <ranges>
-#include <utility>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_copy_if/test.cpp:10]: Technically, we still need `<utility>` for `move()`.

**Comment by StephanTLavavej** [comment]
Thanks for this series of improvements! One might say that this is a total *copycat* of #956. :cat2:

**Pull request merged**